### PR TITLE
MultiServer and clients: async coroutine starter in Utils.py

### PR DESCRIPTION
**Pull request by beauxq**
from discord:

> the codebase uses asyncio.create_task a lot
> https://docs.python.org/3.10/library/asyncio-task.html#asyncio.create_task says
>
> > Important: Save a reference to the result of this function, to avoid a task disappearing mid-execution. The event loop only keeps weak references to tasks. A task that isn\u2019t referenced elsewhere may get garbage collected at any time, even before it\u2019s done.
>
> but more often than not, ap code does not store the created task (51 occurrences in 9 files)
> what gives?

## What is this fixing or adding?

a wrapper and `asyncio.create_task` in Utils.py

## How was this tested?

~~none yet~~

Played a multiworld game with Zillion and SNIClient Super Metroid (and a few other players that were not on this version of the client). It all worked in that game.

### CODE REVIEW

**Review by black-sliver — COMMENTED** [review]
`self.snes_socket.close()` and `snes_autoreconnect(ctx)` i feel like should keep a "local" reference (in ctx), where doing a second disconnect discards the first one (i.e. overwrites the old reference) and the second autoreconnect cancels the first one, and is cancelled during shutdown? not sure.

Also I would prefer @Berserker66 to review this, given that I have very little experience with asyncio.

**Review by recklesscoder — COMMENTED** [review]
I think this is a semantically neutral refactor except for the desired effect of avoiding dangling tasks, and I think it addresses all the offending instances of `asyncio.create_task`.

Considering `class AsyncStarter` does not have any other methods and `AsyncStarter.start` is a bit wordy, I would suggest to perhaps additionally provide a module-level function called `async_start` (or maybe `async_faf` for "fire and forget", to be a bit more explicit?) in Utils, that could just wrap the static class method.

I don't think it is necessary or even advisable to add other changes to this PR, although I agree that the auto-reconnect systems in both CommonClient and SNI have (and had) issues in that the reconnect task is not aborted if a manual connect or disconnect is issued during the reconnect delay and thus may fire later at some seemingly random point in time.

**Comment by beauxq** [comment]
Another thing I considered was no class, and just a module level function and `_fof_tasks` at the module level. I wonder what people think about that.
Having both a class and a function that wraps the class seems like a little much to me.

**Comment by Berserker66** [comment]
Utils is mostly functions, except where we provide subclasses. So, in keeping with that I think a utility function fits best into this type of library.

**Comment by beauxq** [comment]
I did this change from class to function. I haven't tested it yet after this change. I'll try to get to testing it again tomorrow.
It would be nice if others also did a test of some of the clients that were touched.

**Review by recklesscoder — APPROVED** [review]
Looks good to me. Did some mild testing with MultiServer, TextClient and FactorioClient and didn't notice anything out of the ordinary.

File: Utils.py [hunk 1]
```diff
@@ -650,3 +651,21 @@ def read_snes_rom(stream: BinaryIO, strip_header: bool = True) -> bytearray:
     if strip_header and len(buffer) % 0x400 == 0x200:
         return buffer[0x200:]
     return buffer
+
+
+_faf_tasks: "Set[asyncio.Task[None]]" = set()
```
[the remark this answers is not in the document]
Comment (beauxq) [reply Utils.py:656]: The quotes are required for Python 3.8 ("Task is no subscriptable").

[the remark this answers is not in the document]
Comment (recklesscoder) [reply Utils.py:656]: Ooh, I didn't know about that.

File: Utils.py [hunk 2]
```diff
@@ -650,3 +651,21 @@ def read_snes_rom(stream: BinaryIO, strip_header: bool = True) -> bytearray:
...
+
+
+def async_start(co: Coroutine[None, None, None], name: Optional[str] = None) -> None:
+    """ Use this to start an async function running without waiting for it. "fire-and-forget" """
```
[the remark this answers is not in the document]
Comment (Berserker66) [reply Utils.py:660]: "Use this to start a task when you don't keep a reference to it or immediatly await it, to prevent early gargabe collection."?

**Comment by beauxq** [comment]
I went all the way through a self-async Zillion and Super Metroid game with no issues.

**Review by Berserker66 — APPROVED** [review]
